libtenzir/src/tql2/resolve.cpp

```diff
@@ -37,43 +40,83 @@ class entity_resolver : public ast::visitor<entity_resolver> {
+                                      size_t idx, entity_ns ns) {
+      result_ = failure::promise();
+      const auto available = std::invoke([&] {
         switch (ns) {
           case entity_ns::op:
-            return "operator";
+            return reg_.operator_names();
           case entity_ns::fn:
-            return "function";
+            return reg_.function_names();
           case entity_ns::mod:
-            return "module";
+            return reg_.module_names();
         }
         TENZIR_UNREACHABLE();
       });
-      // TODO: This list is too long. Suggest only close matches instead. Also,
-      // we should maybe only suggest things in the same module.
-      auto available = std::invoke([&] {
+      const auto type = std::invoke([&] {
         switch (ns) {
           case entity_ns::op:
-            return reg_.operator_names();
+            return "operator";
           case entity_ns::fn:
-            return reg_.function_names();
+            return "function";
           case entity_ns::mod:
-            return reg_.module_names();
+            return "module";
         }
         TENZIR_UNREACHABLE();
       });
-      auto diag = diagnostic::error("{} `{}` not found", type, ident.name)
-                    .primary(ident);
-      if (not available.empty()) {
-        diag = std::move(diag).hint("must be one of: {}",
-                                    fmt::join(available, ", "));
+      const auto prefix = fmt::format(
+        "{}", fmt::join(std::views::transform(path, &ast::identifier::name)
+                          | std::views::take(idx),
+                        "::"));
+      const auto full = fmt::format(
+        "{}", fmt::join(std::views::transform(path, &ast::identifier::name)
+                          | std::views::take(idx + 1),
+                        "::"));
+      const auto& ident = path.at(idx);
+      if (available.empty()) {
+        diagnostic::error("{} `{}` not found", type, ident.name)
+          .primary(ident)
+          .note("no {}s found", type)
+          .emit(diag_);
+        return;
       }
-      std::move(diag).emit(diag_);
-      result_ = failure::promise();
+      if (ns == entity_ns::mod and idx != 0) {
+        diagnostic::error("{} `{}` not found", type, ident.name)
+          .primary(ident)
+          .docs("https://docs.tenzir.com/tql2/{}",
+                target_ns == entity_ns::op ? "operators" : "functions")
+          .emit(diag_);
+        return;
+      }
+      auto filtered = std::views::filter(available, [&](auto&& x) {
+        return x.starts_with(prefix);
+      });
+      auto best = std::ranges::max(filtered, {}, [&](auto&& x) {
+        return detail::score(full, x);
+      });
+      const auto best_score = detail::score(full, best);
+      if (const auto pos = best.find(prefix); not prefix.empty() and pos == 0) {
+        best.erase(0, prefix.size() + 2);
+      }
+      if (best_score > -5) {
```

Comment:
So this magic number effectively means that... Assuming the candidate is not a continuation of what the user typed, then the edit distance must not be greater than 5, so `x` is a valid suggestion for `quxbar` and vice-versa. If however the candidate is a continuation, then the edit distance must not be greater than `5 - len(user_input)`. I find this a bit weird and maybe it's a result of overfitting the heuristic to the available examples, but whatever. We can just roll with it for now and see how it performs in practice :+1: